include/libtorrent/hash_picker.hpp

```diff
@@ -0,0 +1,217 @@
+
+		hash_request pick_hashes(typed_bitfield<piece_index_t> const& pieces);
+
+		add_hashes_result add_hashes(hash_request const& req, span<sha256_hash const> hashes);
+		// TODO: support batched adding of block hashes for reduced overhead?
+		set_block_hash_result set_block_hash(piece_index_t piece, int offset, sha256_hash const& h);
+		void hashes_rejected(hash_request const& req);
+		void verify_block_hashes(piece_index_t index);
+
+		// do we know the piece layer hash for a piece
+		bool have_hash(piece_index_t index) const;
+		// do we know all the block hashes for a file?
+		bool have_all(file_index_t file) const;
+		bool have_all() const;
+		// get bits indicating if each leaf hash is verified
+		aux::vector<aux::vector<bool>, file_index_t> const& verified_leafs() const
+		{ return m_hash_verified; }
+		bool piece_verified(piece_index_t piece) const;
+
+		int piece_layer() const { return m_piece_layer; }
+
+	private:
+		// returns the number of proof layers needed to verify the node's hash
+		int layers_to_verify(node_index idx) const;
+		int file_num_layers(file_index_t idx) const;
+
+		struct piece_hash_request
+		{
+			time_point last_request = min_time();
+			int num_requests = 0;
+			bool have = false;
+		};
+
+		struct priority_block_request
+		{
+			priority_block_request(file_index_t const f, int const b)
+				: file(f), block(b) {}
+			file_index_t file;
+			int block;
+			int num_requests = 0;
+			bool operator==(priority_block_request const& o) const
+			{ return file == o.file && block == o.block; }
+			bool operator!=(priority_block_request const& o) const
+			{ return !(*this == o); }
+			bool operator<(priority_block_request const& o) const
+			{ return num_requests < o.num_requests; }
+		};
+
+		struct piece_block_request
+		{
+			piece_block_request(file_index_t const f, int const p) : file(f), piece(p) {}
+			file_index_t file;
+			int piece;
+			time_point last_request;
+			int num_requests = 0;
+			bool operator==(piece_block_request const& o) const
+			{ return file == o.file && piece == o.piece; }
+			bool operator!=(piece_block_request const& o) const
+			{ return !(*this == o); }
+			bool operator<(piece_block_request const& o) const
+			{ return num_requests < o.num_requests; }
+		};
+
+		file_storage const& m_files;
+		aux::vector<aux::vector<sha256_hash>, file_index_t>& m_merkle_trees;
+		aux::vector<aux::vector<bool>, file_index_t> m_hash_verified;
+		aux::vector<aux::vector<piece_hash_request>, file_index_t> m_piece_hash_requested;
+		// blocks are only added to this list if there is a time critial block which
+		// has been downloaded but we don't have its hash or if the initial request
+		// for the hash was rejected
+		// this block hash will be requested from every peer possible until the hash
+		// is received
+		// the vector is sorted by the number of requests sent for each block
+		aux::vector<priority_block_request> m_priority_block_requests;
```

Comment:
I wonder if it would make sense to make this a `std::priority_queue<>`. that could easily be a later optimization too though.